pkg/gateway/server/dispatcher/dispatcher.go

```diff
@@ -24,23 +23,46 @@ import (
 )
 
 type Dispatcher struct {
-	openAIAPIKey string
-	invoker      *invoke.Invoker
-	client       kclient.Client
-	lock         *sync.RWMutex
-	urls         map[string]*url.URL
+	invoker *invoke.Invoker
+	client  kclient.Client
+	lock    *sync.RWMutex
+	urls    map[string]*url.URL
 }
 
 func New(invoker *invoke.Invoker, c kclient.Client) *Dispatcher {
 	return &Dispatcher{
-		openAIAPIKey: os.Getenv("OTTO8_OPENAI_MODEL_PROVIDER_API_KEY"),
-		invoker:      invoker,
-		client:       c,
-		lock:         new(sync.RWMutex),
-		urls:         make(map[string]*url.URL),
+		invoker: invoker,
+		client:  c,
+		lock:    new(sync.RWMutex),
+		urls:    make(map[string]*url.URL),
 	}
 }
 
+func (d *Dispatcher) URLForModelProvider(ctx context.Context, namespace, modelProviderName string) (*url.URL, error) {
+	d.lock.RLock()
+	u, ok := d.urls[modelProviderName]
+	d.lock.RUnlock()
+	if ok && (u.Scheme == "https" || engine.IsDaemonRunning(u.String())) {
+		return u, nil
+	}
+
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	u, ok = d.urls[modelProviderName]
+	if ok && (u.Scheme == "https" || engine.IsDaemonRunning(u.String())) {
+		return u, nil
+	}
```

Comment:
We're checking the same thing twice here. Looks like we were doing this before, too. If there's a specific reason, I think it should be added as a comment in the code.

Reply:
Good callout. I'll add a comment.

Reply:
Comments added.